jax/api.py

```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
+    params['out_tree'].store(out_tree)  # linear_util style side effect
+    return out, out_dot
+  ad.primitive_jvps[fun.prim] = custom_transforms_jvp
+
+def defjvp(fun, *jvprules):
+  """Definine JVP rules for each argument separately.
+
+  This function is a convenience wrapper around ``jax.defjvp_all`` for
+  separately defining JVP rules for each of the function's arguments. This
+  convenience wrapper does not provide a mechanism for depending on anything
+  other than the function arguments, so JVP rules defined in this way can't
+  share work with the primal computation or with each other, though those things
+  are possible using ``jax.defjvp_all``. See ``jax.defjvp2`` for a variant that
+  allows for defining rules that depend on the output primal value. See also the
+  ``jax.defjvp_all`` docstring for more detail.
+
+  The signature of each component JVP rule is ``lambda g, *primals: ...`` where
+  ``g`` represents the tangent of the corresponding positional argument and
+  ``*primals`` represents all the primal positional arguments.
+
+  Defining a custom JVP rule also affects the default VJP rule, which is derived
+  from the JVP rule automatically via transposition.
+
+  Args:
+    fun: a custom_transforms function.
+    *jvprules: a sequence of functions or Nones specifying the JVP rule for each
+      corresponding positional argument. When an element is None, it indicates
+      that the Jacobian from the corresponding input to the output is zero.
+
+  Returns:
+    None. A side-effect is that ``fun`` is associated with the JVP rule
+    specified by ``*jvprules``.
+
+  For example:
+
+  >>> @jax.custom_transforms
+  ... def f(x):
+  ...   return np.sin(x ** 2)
+  ...
+  >>> print(f(3.))
+  0.4121185
+  >>> out_primal, out_tangent = jax.jvp(f, (3.,), (2.,))
+  >>> print(out_primal)
+  0.4121185
+  >>> print(out_tangent)
+  -10.933563
+  >>> jax.defjvp(f, lambda g, x: 8. * g)
+  >>> out_primal, out_tangent = jax.jvp(f, (3.,), (2.,))
+  >>> print(out_primal)
+  0.4121185
+  >>> print(out_tangent)
+  16.0
+
+  An example with a function on two arguments:
+
+  >>> @jax.custom_transforms
+  ... def f(x, y):
+  ...   return np.sin(x ** 2 + y)
+  ...
+  >>> print(f(3., 4.))
+  0.42016703
+  >>> out_primal, out_tangent = jax.jvp(f, (3., 4.), (1., 2.))
+  >>> print(out_primal)
+  0.42016703
+  >>> print(out_tangent)
+  7.2595744
+  >>> jax.defjvp(f, None, lambda g, x, y: 8. * g + y)
+  >>> out_primal, out_tangent = jax.jvp(f, (3., 4.), (1., 2.))
+  >>> print(out_primal)
+  0.42016703
+  >>> print(out_tangent)
+  20.0
+  """
+  _check_custom_transforms_type("defjvp", fun)
+  def custom_jvp(primals, tangents):
+    ans = fun(*primals)
+    tangents_out = [rule(t, *primals) for rule, t in zip(jvprules, tangents)
+                    if rule is not None and t is not ad_util.zero]
+    return ans, reduce(ad.add_tangents, tangents_out, ad_util.zero)
+  defjvp_all(fun, custom_jvp)
+
+def defjvp2(fun, *jvprules):
```

Comment:
I'm not sure what 2 refers to, maybe the two non-variadic arguments to the jvprules functions? Could we get rid of defjvp and only have this one?

Reply:
We could just have one, but the most common pattern in our internal code (where we have `ad.defjvp` and `ad.defjvp2` for Primitives) is not to depend on the output. That is, if you look at our JVP rules in lax.py, we almost always use ``defejvp`` but then sometimes use ``defjvp2``. That and having these functions follow the convention of their counterparts in ad.py seem like good reasons.

The 2 doesn't refer to anything, it's just a way to name a variant of a function while keeping the name short. To me it's like a "prime", but we can't add `'` to a name in Python.

I know you might not like that convention, but we already use it elsewhere (to no ill effect so far!), and in any case it doesn't seem like something worth getting stuck on. WDYT?

Reply:
Maybe `defjvp_separate` or  `devjvp_sep` if you really want to save characters?

For what it's worth I also found the name confusing, and wondered what the 2 was about :)

Reply:
I vote for the current name: `defjvp2` by the same reason as Matt.

Reply:
For jvp rules that don't depend on the output, can't you just put a `_` for that arg and ignore it? If the performance etc. is the same, seems pretty lightweight to me. People (including us) can make their own wrappers if it really bothers them.

Reply:
> For what it's worth I also found the name confusing, and wondered what the 2 was about :)

Wondering what a function is and how it works doesn't seem like a bad thing! I would wonder about what all of these functions mean before looking at example uses or their code (or docstring). ("separate" isn't the right suffix here because the difference in question is between taking the output primal as an argument or not.)

> People (including us) can make their own wrappers if it really bothers them.

Yes, that's the idea, but since `defjvp` is the most common pattern, and `defjvp2` is the more general second-most-common pattern, we provide those wrappers for users. After all, both `defjvp` and `defjvp2` are already convenience wrappers around `defjvp_all`.

I think reasonable people can (and do) disagree about personal preferences here. I'm going to keep `defjvp2`.